urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/tools/load/model/BatchUpdate.java

```diff
@@ -0,0 +1,289 @@
+    	return frameEditsMap;
+    }
+    
+    public TreeSet<String> getFrameIDSet() {
+    	return frameIDSet;
+    }
+    
+    public int getLines() {
+    	return lines.size();
+    }
+    
+    public ArrayList<Event> getEventLog() {
+    	return eventLog;
+    }
+    
+    public DefaultListModel<String> getFrameIDsModel() {
+    	DefaultListModel<String> listModel = new DefaultListModel<String>();
+		for (String frameID : getFrameIDSet()) {
+			listModel.addElement(frameID);
+		}
+		return listModel;
+    }
+    
+    public void downloadFrames(JavacycConnection conn) {
+    	progressMonitor = new ProgressMonitor(DefaultController.mainJFrame, "Analyzing Imports...", "", 0, 100);
+    	task = new DownloadFramesTask(conn, progressMonitor, frameIDSet);
+		
+		task.addPropertyChangeListener(new PropertyChangeListener() {
+            @Override
+            public void propertyChange(PropertyChangeEvent e) {
+                if ("progress".equals(e.getPropertyName())) {
+                	progressMonitor.setProgress((Integer) e.getNewValue());
+                	
+                	if (progressMonitor.isCanceled()) {
+                		task.cancel(true);
+                	}
+                	
+                	if (task.isDone() && !task.isCancelled()) {
+                		try {
+							framesFromKB = task.get();
+							progressMonitor.setProgress(0);
+							progressMonitor.close();
+							loadPanel.showPreviewPanel();
+							loadPanel.addAuthorCredits();
+							DefaultController.mainJFrame.getGlassPane().setVisible(false);
+						} catch (InterruptedException e1) {
+							framesFromKB = new ArrayList<Frame>();
+						} catch (ExecutionException e1) {
+							e1.printStackTrace();
+						}
+                	}
+                	
+                	if (task.isCancelled()) {
+                		framesFromKB = new ArrayList<Frame>();
+                		progressMonitor.setProgress(0);
+						progressMonitor.close();
+						DefaultController.mainJFrame.getGlassPane().setVisible(false);
+                	}
+                }
+            }
+        });
+		DefaultController.mainJFrame.setGlassPane(new TranslucentGlassPane());
+    	DefaultController.mainJFrame.getGlassPane().setVisible(true);
+		task.execute();
+    }
+    
+    public DefaultListModel<String> framesWhichModifyKB(JavacycConnection conn) {
+    	DefaultListModel<String> listModel = new DefaultListModel<String>();
+    	TreeSet<String> frameSet = new TreeSet<String>();
+    	for (AbstractFrameEdit frameEdit : frameEdits) {
+    		Frame frame = getFrameByID(frameEdit.getFrameID());
+    		if (frame == null) frameSet.add(frameEdit.getFrameID());
+    		else {
+	    		boolean modifiesKB = true;
+				try {
+					modifiesKB = frameEdit.modifiesFrame(conn, frame);
+				} catch (PtoolsErrorException e) {
+					e.printStackTrace();
+				}
+	    		if (modifiesKB) {
+	    			frameSet.add(frameEdit.getFrameID());
+	    		}
+    		}
+    	}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 145:159, 167:181.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.